keps/sig-instrumentation/647-apiserver-tracing/README.md

```diff
@@ -0,0 +1,226 @@
+
+```golang
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// OpenTelemetryClientConfiguration provides versioned configuration for opentelemetry clients.
+type OpenTelemetryClientConfiguration struct {
+  metav1.TypeMeta `json:",inline"`
+
+  // +optional
+  // URL of the collector that's running on the master.
+  // if URL is specified, APIServer uses the egressType Master when sending tracing data to the collector.
+  URL *string `json:"url,omitempty" protobuf:"bytes,3,opt,name=url"`
+
+  // +optional
+  // Service that's the frontend of the collector deployment running in the cluster.
+  // If Service is specified, APIServer uses the egressType Cluster when sending tracing data to the collector.
+  Service *ServiceReference `json:"service,omitempty" protobuf:"bytes,1,opt,name=service"`
+}
+
+// ServiceReference holds a reference to Service.legacy.k8s.io
+type ServiceReference struct {
+  // `namespace` is the namespace of the service.
+  // Required
+  Namespace string `json:"namespace" protobuf:"bytes,1,opt,name=namespace"`
+  // `name` is the name of the service.
+  // Required
+  Name string `json:"name" protobuf:"bytes,2,opt,name=name"`
+
+  // If specified, the port on the service.
+  // Defaults to 55680.
+  // `port` should be a valid port number (1-65535, inclusive).
+  // +optional
+  Port *int32 `json:"port,omitempty" protobuf:"varint,3,opt,name=port"`
+}
+```
+
+### Controlling use of the OpenTelemetry library
+
+As the community found in the [Metrics Stability Framework KEP](https://github.com/kubernetes/enhancements/blob/master/keps/sig-instrumentation/20190404-kubernetes-control-plane-metrics-stability.md#kubernetes-control-plane-metrics-stability), having control over how the client libraries are used in kubernetes can enable maintainers to enforce policy and make broad improvements to the quality of telemetry.  To enable future improvements to tracing, we will restrict the direct use of the OpenTelemetry library within the kubernetes code base, and provide wrapped versions of functions we wish to expose in a utility library.
+
+### Test Plan
+
+We will e2e test this feature by deploying an OpenTelemetry Collector on the master, and configure it to export traces using the [stdout exporter](https://github.com/open-telemetry/opentelemetry-go/tree/master/exporters/stdout), which logs the spans in json format.  We can then verify that the logs contain our expected traces when making calls to the API Server.
```

Comment:
Do you plan this to ever be conformance tested? This doesn't sound easy to conformance test.

Reply:
Yes, it would be good to have this conformance tested.  I added it as a GA requirement.  With the current configuration options, we can run the OTel Collector as a service in the cluster instead of on the master node.

Reply:
The thing is, users likely won't be able to configure that config file. But yeah, I agree with testing via running this in the cluster.